Allow gateway exec-ing into a failed solve with an exec op

hinshun opened the pull request:
For: https://github.com/moby/buildkit/issues/1472#issuecomment-705253089

- Adds `Evaluate bool` to the `SolveRequest`, this explicitly evaluates `ResultProxy` because they are currently lazy results (until returned, or a call like `ref.ReadFile` is called). Otherwise, exec errors will only show up in the main solve request instead of inside `runGatewayCB`.
- Adds `SolveError` as a `TypedErrorProto`. Clients can use `errors.As` to extract it.
- Adds `GetDefault() (worker.Worker, nil)` to `frontend.WorkerInfos`, that plumbs down the default worker.
- Refactors `frontend.WorkerInfos` to `worker.Infos` because of the cyclic dependency
- Refactors `gateway.NewContainer` to share the exact same code as `solver/llbsolver/ops/exec.go`, previously was 90% duplicated. The existing version of `gateway.NewContainer` also didn't allow for scratch mounts whereas `solver/llbsolver/ops/exec.go` did.

Example client code: https://gist.github.com/hinshun/e72f509121e022bc81ebba03fc2851c6

When an `pb.ExecOp` or ` pb.FileOp` fails to be solved, its protobuf definition, along with its solved inputs / outputs are wrapped in a typed error to the `lbfBridgeForwarder` which holds the temporary IDs for references.

The inputs and outputs have temporary uuids generated and sent back over as a `TypedErrorProto` which allows it to unmarshaled on the client side to reconstruct the arguments necessary to gateway exec back into a container process for the failed solve.

hinshun commented:
@tonistiigi @ktock Unsure if this is related to my change?
```
    --- FAIL: TestIntegration/TestSecurityModeSysfs/worker=containerd-snapshotter-stargz/secmode=sandbox (1.59s)
        client_test.go:648: 
            	Error Trace:	client_test.go:648
            	            				run.go:171
            	Error:      	"failed to solve: rpc error: code = Unknown desc = exit code: 1" does not contain "executor failed running"
```

coryb commented on frontend/gateway/forwarder/forward.go:
```diff
@@ -161,20 +162,31 @@ func (c *bridgeClient) NewContainer(ctx context.Context, req client.NewContainer
-		var refProxy solver.ResultProxy
+		var workerRef *worker.WorkerRef
+		// TODO: What about ResultID?
+		// The *bridgeClient doesn't have a workerRefByID map.
```
I am not sure this logic is valid, but you might be able to compare the ResultID against all refs in the `brdigeClient`, something like:
```
		if m.ResultID != "" {
			for _, ref := range c.refs {
				im, _ := ref.getImmutableRef(ctx)
				if m.ResultID == im.ID() {
					rr, _ := ref.ResultProxy.Result(ctx)
					workerRef, _ = rr.Sys().(*worker.WorkerRef)
					break
				}
			}
		}
```

hinshun replied:
They don't share the same map unfortunately because with solve errors we are dealing with non-cached `solver.Result`.
I ended up resolving the same way I did in `llbBridgeForwarder`: https://github.com/moby/buildkit/pull/1732/commits/a799187d4f1e69758db97c6b7a0dae46057772b5

hinshun commented on client/build_test.go:
```diff
@@ -950,6 +952,157 @@ func testClientGatewayContainerExecTty(t *testing.T, sb integration.Sandbox) {
+		require.True(t, ok)
+
+		file := solveFile.File
+		require.Len(t, file.Actions, 1)
```
Maybe the `pb.FileOp` returned should only contain the action that failed?

tonistiigi replied:
Then we would need to filter the inputs/outputs as well. And this gets a bit tricky as actions input might be the output of another action. Maybe just another typed indicator that fileop sets. And extra type for a "mount  error"

hinshun replied:
> Maybe just another typed indicator that fileop sets.
Can you elaborate on this idea?

tonistiigi commented:
@hinshun Is this ready?

hinshun commented:
@tonistiigi Yes, though I did punt on the errors before `Run` for `ExecOp`. I don't think its worth the complexity for `SolveError`. I think they can also derive the op from the `VertexError` if they wanted metadata.

tonistiigi commented:
VertexError only contains digest though, not the proto definition.

hinshun commented:
Yeah, I understand clients doesn't always have access to the proto definition to lookup the op via digest. Maybe introduce an `OpError` in the future?

tonistiigi commented:
What is the difference between `OpError` and `SolveError`? I think returning the input refs would be useful for `OpError` as well.

hinshun commented:
@tonistiigi I rebased and added a commit that returns exec errors for errors returned before the executor run now: https://github.com/moby/buildkit/pull/1732/commits/1bbc7c203032cc6460d4b6f876d750ad283af87e

coryb commented on frontend/gateway/forwarder/forward.go:
```diff
@@ -113,6 +118,59 @@ func (c *bridgeClient) Inputs(ctx context.Context) (map[string]llb.State, error)
+		var err error
+		inputIDs, err = c.registerResultIDs(ee.Inputs...)
+		if err != nil {
+			return err
```
In the edge case that the `registerResultIDs` returns the `unexpected type for result` error then we will lose the original solve error here.  I wonder if these should be `errors.Wrap(solveErr, err.Error())` instead to make sure the solve error is preserved?  I am not sure if we would ever practically get the `unexpected type for result` error though.

hinshun replied:
> I am not sure if we would ever practically get the unexpected type for result error though.
I think that would be an implementation error in Buildkit?

I'm okay with the `errors.Wrap` approach, but unsure if it'll be useful if incomplete? You'll need a lot of safeguards in the client side for incomplete solve errors.

coryb commented on frontend/gateway/gateway.go:
```diff
@@ -486,6 +493,59 @@ func translateLegacySolveRequest(req *pb.SolveRequest) error {
+		var err error
+		inputIDs, err = lbf.registerResultIDs(ee.Inputs...)
+		if err != nil {
+			return err
```
same issue here, we will lose solveErr

coryb commented on solver/llbsolver/solver.go:
```diff
@@ -142,6 +143,20 @@ func (s *Solver) Solve(ctx context.Context, id string, sessionID string, req fro
+
+						workerRef, ok := res.Sys().(*worker.WorkerRef)
+						if !ok {
+							return nil
```
Why are we returning `nil` here?  It seems like we should return `err` from the `ref.Result`?

hinshun replied:
The `err` from `ref.Result` is returned on line 160, and we're currently dropping the error from `ee.EachRef` like how we do in the defer on line 135 when dealing with releasing results, but perhaps I shouldn't be following that anti-pattern?

coryb replied:
I see, I missed this was in the `EachRef` callback, it seems fine as it.

tonistiigi commented on frontend/gateway/forwarder/forward.go:
```diff
@@ -161,20 +195,35 @@ func (c *bridgeClient) NewContainer(ctx context.Context, req client.NewContainer
 				return nil, errors.Errorf("unexpected Ref type: %T", m.Ref)
 			}
+
+			res, err := refProxy.Result(ctx)
```
^

hinshun replied:
Done here: https://github.com/moby/buildkit/pull/1732/commits/dfaf61399675def42b29bfcd0eab220a982461bc

hinshun commented on solver/llbsolver/ops/file.go:
```diff
@@ -397,18 +400,42 @@ func (s *FileOpSolver) getInput(ctx context.Context, idx int, inputs []fileoptyp
+					}
+				}
+
+				err = errdefs.WithExecError(err, inputRes, outputRes)
```
`Exec` as opposed to `CacheMap`, not the `ExecOp`. I'm not sure what's a better name... maybe `ResultError`?

tonistiigi replied:
Ok, I guess we don't need to block on that. It probably makes sense to rename the `Exec` function in `Op` as well to avoid confusion but that can be follow-up.

hinshun commented on client/build_test.go:
```diff
@@ -1069,6 +1073,339 @@ func testClientGatewayContainerPlatformPATH(t *testing.T, sb integration.Sandbox
+			mounts = append(mounts, client.Mount{
+				Selector:  mnt.Selector,
+				Dest:      mnt.Dest,
+				ResultID:  se.Solve.OutputIDs[mnt.Output],
```
Both work, they could use `mnt.Input` but they would be the unmodified inputs. For this test, I was checking that I could access the mutated mounts rather than the original inputs.

In the LLB we have: `echo %s > output && fail`, so this test was to exec into a mount where `echo %s > output` succeeded but the `fail` failed the execop overall.

tonistiigi replied:
The logic should be that both `se.Solve.OutputIDs` and `se.Solve.InputIDs` are indexed by `mnt.Input`. `mnt.Output` is a link to the next vertex. There is no need that output must be defined on a mount in order to be able to debug its error state.

tonistiigi commented:
@hinshun Any update? Mainly on the ref indexing that I think needs changes as we discussed in slack. If we can get this over the line we could do `v0.8rc`

hinshun commented:
@tonistiigi Sorry have been on my week long oncall at work. I don't expect the ref indexing take too long to implement, but I think I'll need to clarify some of the edge cases with you tomorrow.

tonistiigi commented on frontend/gateway/container.go:
```diff
@@ -71,122 +53,202 @@ func NewContainer(ctx context.Context, e executor.Executor, sm *session.Manager,
+			}
+			mountable = active
+			p.Actives = append(p.Actives, active)
+			if m.Output != opspb.SkipOutput && ref != nil {
```
I wonder if having output in here is even supported. Maybe just error?

tonistiigi replied:
At least client protects against this https://github.com/moby/buildkit/blob/c70058095bd131bfbdca4d0796a2feb042bd1de8/client/llb/exec.go#L307

hinshun replied:
Hmm I haven't changed this conditional, it's the same before refactoring here:
https://github.com/moby/buildkit/blob/9369d536788e44b8b991745a68bd9966e0239141/solver/llbsolver/ops/exec.go#L294

tonistiigi replied:
Ok, yeah I missed that input is cloned to be output here. Still a weird case but no changes needed for this PR.

tonistiigi commented on frontend/gateway/forwarder/forward.go:
```diff
@@ -247,19 +246,26 @@ func (c *bridgeClient) NewContainer(ctx context.Context, req client.NewContainer
 		})
 	}
 
+	w, err := c.workers.GetDefault()
```
We can leave this to follow-up but this looks logically wrong to have worker resolving in this level. `GetDefault()` is a function could be replaced with a more complex worker resolve function in the future but logic like that shouldn't be in this package but figured out in `llbsolver`

hinshun replied:
Sounds good. Since the `worker.Info` is just an interface, we can pass an implementation that returns the specific worker resolved by `llbsolver` in the future. Either way, the downstream consumers (like the gateway forwarder) needs access to a `worker.Worker` and `[]client.WorkerInfo`, so we can just rename `GetDefault` -> `GetResolvedWorker` or something.

tonistiigi commented on solver/llbsolver/ops/exec.go:
```diff
@@ -214,149 +214,48 @@ func addDefaultEnvvar(env []string, k, v string) []string {
-			if mountable == nil {
-				continue
+			execOutputs := make([]solver.Result, len(e.op.Mounts))
+			for i, res := range results {
```
I still don't quite understand how this works. We are mapping only mounts that had set output index, instead of all the mounts(that would then point to either mutable or same input if readonly). I think the test works because something always fills up output index there, even if it is unused.

hinshun replied:
Addressed here: https://github.com/moby/buildkit/pull/1732/commits/1240dd7795823218c049782ce0302776a69e0739

tonistiigi commented on client/build_test.go:
```diff
@@ -1084,79 +1084,139 @@ func testClientGatewayExecError(t *testing.T, sb integration.Sandbox) {
+			"rootfs and readwrite mount",
+			llb.Image("busybox:latest").Run(
+				llb.Shlexf(`sh -c "echo %s > /data && echo %s > /rw/data && fail"`, id, id),
+				llb.AddMount("/rw", llb.Scratch().File(llb.Mkfile("foo", 0700, []byte(id)))),
```
If you would set `llb.ForceNoOutput()` in here I think this test would not work.

hinshun replied:
Confirmed it doesn't work, thanks. The `MountID[1]` becomes scratch when setting `llb.ForceNoOutput`.

hinshun commented on frontend/gateway/container.go:
```diff
@@ -71,122 +53,202 @@ func NewContainer(ctx context.Context, e executor.Executor, sm *session.Manager,
-				}
+		// if mount is based on input validate and load it
+		if m.Input != opspb.Empty {
+			if int(m.Input) > len(refs) {
```
Fixed now, but note this is an existing bug: https://github.com/moby/buildkit/blob/9369d536788e44b8b991745a68bd9966e0239141/solver/llbsolver/ops/exec.go#L242

hinshun commented:
PR comments have been addressed.
Note that I changed when we wrap the `OpError`: https://github.com/moby/buildkit/pull/1732/commits/1240dd7795823218c049782ce0302776a69e0739#diff-2fc2e99eea0899fac8a14da1878d67aa8520bed09be4de598226d4d29450d6c8L201-L215

I noticed there were edge cases of some kind that the digest from `VertexError` doesn't correspond to any op inside the `*pb.Definition`. I took the safe route of wrapping the op where I know it will have the correct one in `solver/jobs.go`, by doing the cast `s.st.vtx.Sys().(*pb.Op)` and allowing me to pass all the tests.

coryb reviewed (approved):
Looks great! Sorry I missed the scratch mounting via NewContainer, thanks for fixing that.
